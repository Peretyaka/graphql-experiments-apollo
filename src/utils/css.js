import {TRANSITION_DURATION, TIMING_FUNCTION} from '../constants/animations'

/**
 * Border mixin
 * @param {string} border color
 */
export const border = (borderColor = '#000') => ({
  width: '1px', // jss-default-unit and jss-expand bug. Deep passing interprets 'width' as not a part of 'border' statement
  style: 'solid',
  color: borderColor,
})

/**
 * Transition mixin
 * @param {number} transition
 * @param {number} delay
 */
export const transition = (duration = `${TRANSITION_DURATION}ms`, delay = '0ms') => ({
  property: 'all',
  timingFunction: TIMING_FUNCTION,
  duration,
  delay,
})

export default {
  border,
  transition
}
