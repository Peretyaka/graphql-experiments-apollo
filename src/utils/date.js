/**
 * Convert unix time to something more readable
 * @param {number} timestamp
 */
export const getUserReadableDate = timestamp => (new Date(timestamp)).toDateString()

export default {
  getUserReadableDate
}
