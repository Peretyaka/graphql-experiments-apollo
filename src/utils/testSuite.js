import React from 'react'
import PropTypes from 'prop-types'
import {create as createJss} from 'jss'
import preset from 'jss-preset-default'
import {JssProvider, ThemeProvider} from 'react-jss'
import Enzyme, {
  render as enzymeRender,
  shallow as enzymeShallow,
  mount as enzymeMount
} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import theme from '../theme'

// Setup enzyme adapter for current version of React
Enzyme.configure({
  adapter: new Adapter()
})

/**
 * Create testing jss instance
 * NOTE: For simplifying testing. Just return original classname
 */
const jss = createJss({
  createGenerateClassName: () => rule => rule.key,
  ...preset(),
})

/**
 * Create helper component on top of enzyme render
 * methods due to fact, that we are using JSS and it
 * can't work without initialized of "theme" and "jss"
 * providers in root from application.
 */
const WrappedWithJss = ({children}) => (
  <ThemeProvider theme={theme}>
    <JssProvider jss={jss}>
      {children}
    </JssProvider>
  </ThemeProvider>
)

WrappedWithJss.propTypes = {
  children: PropTypes.node.isRequired
}

export const shallowWithJss = children => enzymeShallow(
  <WrappedWithJss>{children}</WrappedWithJss>
)

export const renderWithJss = children => enzymeRender(
  <WrappedWithJss>{children}</WrappedWithJss>
)

export const mountWithJss = children => enzymeMount(
  <WrappedWithJss>{children}</WrappedWithJss>
)

/**
 * Helper method for 'with JSS' render methods above.
 * Extracts from deep nested tree needed component.
 *
 * All this nesting is (from above down to component):
 * 1. ThemeProvider
 * 2. JssProvider
 * 3. Jss
 * 4. Our needed component
 */

export const extractFromJssWrapper = wrap => wrap.children().children().children().children()

/**
 * Proxy methods to give access to enzyme render methods from
 * one place. Also, they are created to avoid initalizing
 * in all test files enzyme adapters
 */

export const shallow = children => enzymeShallow(children)

export const render = children => enzymeRender(children)

export const mount = children => enzymeMount(children)
