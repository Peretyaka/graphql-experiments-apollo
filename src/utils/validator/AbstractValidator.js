export default class AbstractValidator {
  static defaults = {
    message: '',
  }

  constructor(options = {}) {
    this.options = Object.assign({}, this.defaults, options)
  }

  /**
   * @public
   * @param {Object} data
   * @returns {Promise}
   */
  validate(value) {
    return new Promise((resolve, reject) => {
      if (this.validateValue(value)) resolve()
      reject(this.options.message)
    })
  }

  /**
   * @abstract
   */
  // eslint-disable-next-line
  validateValue() {
    return true
  }
}
