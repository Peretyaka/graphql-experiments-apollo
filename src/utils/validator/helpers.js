/**
 * Extracts possible error from error object
 * @param {Object} errors
 * @param {string} field
 * @returns {null|string}
 */
export const getErrorByField = (errors, field) => errors && errors[field] && errors[field][0]

export default {
  getErrorByField
}
