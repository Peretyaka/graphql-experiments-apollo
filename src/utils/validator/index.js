export {default as validate} from './validate'
export {getErrorByField} from './helpers'
export {default as RequiredValidator} from './RequiredValidator'
export {default as StringValidator} from './StringValidator'
