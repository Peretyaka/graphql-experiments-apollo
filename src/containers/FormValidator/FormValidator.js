import {PureComponent} from 'react'
import PropTypes from 'prop-types'
import {validate} from '../../utils/validator'

/**
 * Form fields validator.
 * As rules recieves an object like this:
 *
 * {
 *     firstField: [
 *         new StringValidator({
 *             message: 'You provide wrong field',
 *             min: 5,
 *             max: 20
 *         })
 *     ],
 *     secondField: [
 *         new EmailValidator({
 *             message: 'Invalid email'
 *         }),
 *         new RequiredValidator({
 *             message: 'This field is required'
 *         })
 *     ]
 * }
 *
 * Example usage:
 *
 * import {RequiredValidator, getErrorByField} from '../../utils/validator'
 * import FormValidator from '../../containers/FormValidator'
 *
 * export default () => (
 *   <FormValidator
 *     data={{foo: 'bar'}}
 *     rules={{foo: [new RequiredValidator({message: 'Foo should not be empty'})]}}
 *   >
 *     {({success, errors, data}) => {
 *       if (errors) {
 *         return <div>{getErrorByField(errors, 'foo')}</div>
 *       }
 *       return <div>All OK</div>
 *     }}
 *   </FormValidator>
 * )
 *
 * Returns next params:
 * 1. 'success' - boolean
 * 2. 'error' - null or object of errors
 * 3. 'data' - original data
 */
class FormValidator extends PureComponent {

  static propTypes = {
    data: PropTypes.objectOf(PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number
    ])),
    rules: PropTypes.objectOf(PropTypes.arrayOf(PropTypes.object)).isRequired,
    children: PropTypes.func.isRequired,
  }

  constructor(props) {
    super(props)
    this.state = {
      /**
       * Will be true if form is successfully validated
       * @type {boolean}
       */
      success: null,

      /**
       * Will be true if there are any validation error
       * @type {boolean}
       */
      errors: null,
    }

    if (!props.rules) {
      console.error('FormValidator: no rules specified. Usage of this component doesn\'t make sense.') // eslint-disable-line
    }
  }

  /**
   * Do all validation on props update
   */
  componentDidUpdate(prevProps) {
    const {data, rules} = this.props

    if (prevProps.data === data || !rules) {
      return
    }

    validate(data, rules)
      .then(() => this.setState({
        success: true,
        errors: null,
      }))
      .catch(errors => this.setState({
        success: false,
        errors,
      }))
  }

  render() {
    const {data} = this.props
    const {success, errors} = this.state

    return this.props.children({
      success,
      errors,
      data,
    })
  }
}

export default FormValidator
