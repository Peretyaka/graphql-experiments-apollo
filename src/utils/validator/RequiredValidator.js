import AbstractValidator from './AbstractValidator'

export default class RequiredValidator extends AbstractValidator {

  static defaults = {
    message: '',
  }

  /**
   * @param {string} value
   * @returns {bool}
   */
  validateValue(value) { // eslint-disable-line
    return !!value
  }
}
