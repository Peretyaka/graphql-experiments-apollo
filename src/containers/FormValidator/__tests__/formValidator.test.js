import React from 'react'
import FormValidator from '../FormValidator'
import {RequiredValidator} from '../../../utils/validator'
import {mount} from '../../../utils/testSuite'

const renderValidator = (value) =>
  mount(
    <FormValidator
      data={{foo: value}}
      rules={{
        foo: [new RequiredValidator({message: 'Foo should not be empty'})],
      }}
    >
      {({success, errors}) => {
        if (errors) return <div>Error</div>
        if (success) return <div>Success</div>
        return <div>Default behaviour</div>
      }}
    </FormValidator>,
  )

describe('FormValidator', () => {
  it('Should render correcty in case of successfull validation', (done) => {
    const wrap = renderValidator('foo')

    // Imitate props changing from outside
    wrap.instance().componentDidUpdate({})

    setTimeout(() => {
      wrap.update() // Force re-render after imitating props update
      expect(wrap).toMatchSnapshot()
      done()
    }, 5)
  })

  it('Should render correcty in case of error validation', (done) => {
    const wrap = renderValidator('')

    // Imitate props changing from outside
    wrap.instance().componentDidUpdate({})

    setTimeout(() => {
      wrap.update() // Force re-render after imitating props update
      expect(wrap).toMatchSnapshot()
      done()
    }, 5)
  })

  it("Should do nothing if props wasn't changed (do nothing on init)", () => {
    const wrap = renderValidator('foo')
    expect(wrap).toMatchSnapshot()
  })

  it('Should return success, error and data as callback params', () => {
    const inputData = {foo: 'bar'}
    const wrap = mount(
      <FormValidator data={inputData} rules={{}}>
        {({success, errors, data}) => (
          <div>
            {`Success: ${success}`}
            {`Errors: ${errors}`}
            {`Input data doesn't mutate: ${data === inputData}`}
          </div>
        )}
      </FormValidator>,
    )

    wrap.setState({
      success: 'foo',
      errors: 'bar',
    })

    wrap.update()
    expect(wrap).toMatchSnapshot()
  })
})
