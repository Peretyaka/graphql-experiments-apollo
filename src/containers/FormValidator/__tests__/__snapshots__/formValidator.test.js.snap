// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormValidator Should do nothing if props wasn't changed (do nothing on init) 1`] = `
<FormValidator
  data={
    Object {
      "foo": "foo",
    }
  }
  rules={
    Object {
      "foo": Array [
        RequiredValidator {
          "options": Object {
            "message": "Foo should not be empty",
          },
        },
      ],
    }
  }
>
  <div>
    Default behaviour
  </div>
</FormValidator>
`;

exports[`FormValidator Should render correcty in case of error validation 1`] = `
<FormValidator
  data={
    Object {
      "foo": "",
    }
  }
  rules={
    Object {
      "foo": Array [
        RequiredValidator {
          "options": Object {
            "message": "Foo should not be empty",
          },
        },
      ],
    }
  }
>
  <div>
    Error
  </div>
</FormValidator>
`;

exports[`FormValidator Should render correcty in case of successfull validation 1`] = `
<FormValidator
  data={
    Object {
      "foo": "foo",
    }
  }
  rules={
    Object {
      "foo": Array [
        RequiredValidator {
          "options": Object {
            "message": "Foo should not be empty",
          },
        },
      ],
    }
  }
>
  <div>
    Success
  </div>
</FormValidator>
`;

exports[`FormValidator Should return success, error and data as callback params 1`] = `
<FormValidator
  data={
    Object {
      "foo": "bar",
    }
  }
  rules={Object {}}
>
  <div>
    Success: foo
    Errors: bar
    Input data doesn't mutate: true
  </div>
</FormValidator>
`;
